CDCgov/prime-reportstream
# jimfuqian/9285 Add OpenAPI spec generation task and generate spec for ApiKeysFunction.kt

This PR:

Serves as a pilot (initial step) for generating OpenAPI (https://swagger.io/specification/) doc
for Report Stream API.

Quote User Story:

"As a consumer of the report stream API, I want the documentation of the public key endpoints to be automatically kept in sync with the code"

This PR contains

1. Gradle task "generateOpenApi" - can be run stand alone, from CLI or within IDE Gradle plugin
2. ApiKeysFunctions.kt - Added swagger annotations for all end points

Test Steps:

1. Check out the PR to local repo
3. Run: cd <local repo base dir>/prime-router/
4. Run: ./.gradlew generateOpenApi
5. Sample output, see below:
`./gradlew generateOpenApi

> Configure project :
default messager has already been registered

> Task :generateOpenApi
SLF4J: No SLF4J providers were found.
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See https://www.slf4j.org/codes.html#noProviders for further details.
SLF4J: Class path contains SLF4J bindings targeting slf4j-api versions 1.7.x or earlier.
SLF4J: Ignoring binding found at [jar:file:/home/jfu/.gradle/caches/modules-2/files-2.1/org.apache.logging.log4j/log4j-slf4j-impl/3.0.0-alpha1/e487d0eeb91c64515db130475110376c65ab0634/log4j-slf4j-impl-3.0.0-alpha1.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See https://www.slf4j.org/codes.html#ignoredBindings for an explanation.

BUILD SUCCESSFUL in 5s
6 actionable tasks: 2 executed, 4 up-to-date`

Check the generated OpenApi spec (in yaml): 

ls ./docs/api/generated/api.yaml

## Changes
- *See Files changed*

## Checklist

### Testing
- [x] Run ./gradlew generateOpenApi and observed the yaml generated at ./docs/api/generated/api.yaml
- [x] To get a visual swagger UI page with the api.yaml, go to swagger UI editor at https://editor.swagger.io/, choose Import File, and navigate to the api.yaml generated on your local file system, and the swagger UI editor will render it.

![image](https://github.com/CDCgov/prime-reportstream/assets/135686833/2364d428-5531-4b0c-b75c-5dbd922dc87e)

### Attached Files:

### Process

## Linked Issues

- Fixes #9285 

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #issue
- Ticketed a research on OpenAPI spec generation from bytecode (pros vs cons): see #10104 

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?
